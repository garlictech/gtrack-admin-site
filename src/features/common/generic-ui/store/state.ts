export interface State {
  progressSpinnerOn: boolean;
  progressSpinnerText?: string;
}

export const featureName = 'features.generic-ui';
