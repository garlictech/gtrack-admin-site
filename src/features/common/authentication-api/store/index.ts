import * as Actions from './actions';

export type Action = Actions.AllActions;
export { Effects } from './effects';

export { AuthenticationState, Auth, domain } from './state';
export { reducer as Reducer } from './reducer';
export { Actions };
