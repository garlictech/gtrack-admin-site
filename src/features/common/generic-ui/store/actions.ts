// tslint:disable:no-property-initializers max-classes-per-file
import { Action } from '@ngrx/store';
import { Toast } from '../interfaces';

export enum ActionTypes {
  ShowProgressSpinner = '[UI] Show progress spinner',
  HideProgressSpinner = '[UI] Hide progress spinner',
  DisplayToast = '[UI] display toast'
}

/**
 * The action displays a progress spinner hiding the full screen.
 *
 * @export
 */
export class ShowProgressSpinner implements Action {
  readonly type = ActionTypes.ShowProgressSpinner;
  /**
   * Creates an instance of ShowProgressSpinner.
   * @param [textLabel] A translateable, optional label that appears below the spinner. Use to add some
   * extra info about what is going on.
   */
  constructor(public textLabel?: string) {
    // EMPTY
  }
}

/**
 * Hide the progress spinner displeyed with {@link ShowProgressSpinner}.
 *
 * @export
 */
export class HideProgressSpinner implements Action {
  readonly type = ActionTypes.HideProgressSpinner;
}

/**
 * Display a toast message.
 *
 * @export
 */
export class DisplayToast implements Action {
  readonly type = ActionTypes.DisplayToast;
  /**
   * Creates an instance of DisplayToast.
   * @param notification The toast descriptor.
   */
  constructor(public notification: Toast) {
    // EMPTY
  }
}

export type AllActions = ShowProgressSpinner | HideProgressSpinner | DisplayToast;
