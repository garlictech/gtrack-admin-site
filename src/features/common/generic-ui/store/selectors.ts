import { createFeatureSelector, createSelector } from '@ngrx/store';

import { featureName, State } from './state';

export const selectFeature = createFeatureSelector<State>(featureName);

export const selectProgressSpinnerOn = createSelector(
  selectFeature,
  (state: State) => state.progressSpinnerOn
);
export const selectProgressSpinnerText = createSelector(
  selectFeature,
  (state: State) => state.progressSpinnerText
);
