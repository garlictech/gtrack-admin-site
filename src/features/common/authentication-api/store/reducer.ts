// tslint:disable:only-arrow-functions no-small-switch
import { ActionReducer } from '@ngrx/store';

import * as ApiActions from './actions';
import { AuthenticationState } from './state';

export type Action = ApiActions.AllActions;

const initialState: AuthenticationState = {
  auth: {
    token: undefined,
    refreshToken: undefined,
    user: undefined
  },
  loggingIn: true,
  failed: undefined,
  emailSent: false
};

export const reducer: ActionReducer<AuthenticationState> = (state = initialState, action: Action) => {
  switch (action.type) {
    case ApiActions.TWITTER_LOGIN:
    case ApiActions.REQUEST_VERIFY_TOKEN:
    case ApiActions.FACEBOOK_LOGIN:
    case ApiActions.GOOGLE_LOGIN:
    case ApiActions.MAGICLINK_REQUEST_TOKEN:
    case ApiActions.VERIFY:
    case ApiActions.MAGICLINK_LOGIN:
      return { ...state, loggingIn: true, failed: undefined };

    case ApiActions.MAGICLINK_EMAIL_SENT:
      return { ...state, emailSent: true, failed: undefined, loggingIn: false };

    case ApiActions.LOGIN_SUCCESS:
      return { ...{ auth: action.auth }, loggingIn: false, failed: undefined, emailSent: false };

    case ApiActions.USER_CANCELLED:
    case ApiActions.FAILURE_HAPPENED:
      return { ...state, ...{ failed: action.error }, loggingIn: false, emailSent: false };

    case ApiActions.LOGOUT_START:
      return { ...state };

    case ApiActions.LOGOUT_SUCCESS:
      return { ...initialState, loggingIn: false };

    default:
      return state;
  }
};
