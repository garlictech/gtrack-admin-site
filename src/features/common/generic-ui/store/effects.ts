// tslint:disable:no-property-initializers
import { Inject, Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { TranslateService } from '@ngx-translate/core';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { EToastSeverity, GENERIC_UI_PLATFORM_SERVICE, GenericUiPlatformServiceData } from '../interfaces';
import * as fromActions from './actions';

@Injectable()
export class Effects {
  @Effect({ dispatch: false }) showProgressSpinner$: Observable<Action> = this.actions$.pipe(
    ofType<fromActions.ShowProgressSpinner>(fromActions.ActionTypes.ShowProgressSpinner),
    tap((action: fromActions.ShowProgressSpinner) => this._ps.displayLoader(action.textLabel))
  );

  @Effect({ dispatch: false }) hideProgressSpinner$: Observable<Action> = this.actions$.pipe(
    ofType(fromActions.ActionTypes.HideProgressSpinner),
    tap(() => this._ps.dismissLoader())
  );

  @Effect({ dispatch: false }) displayToast$: Observable<Action> = this.actions$.pipe(
    ofType<fromActions.DisplayToast>(fromActions.ActionTypes.DisplayToast),
    tap((action: fromActions.DisplayToast) =>
      this._ps.displayToast({
        severity: action.notification.severity || EToastSeverity.Success,
        summary: this._translate.instant(action.notification.summary),
        detail: action.notification.detail ? this._translate.instant(action.notification.detail) : undefined
      })
    )
  );
  constructor(
    private readonly actions$: Actions,
    @Inject(GENERIC_UI_PLATFORM_SERVICE) private readonly _ps: GenericUiPlatformServiceData,
    private readonly _translate: TranslateService
  ) {
    // EMPTY
  }
}
