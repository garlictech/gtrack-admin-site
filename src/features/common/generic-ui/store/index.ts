import * as Actions from './actions';

export type Action = Actions.AllActions;
export { Effects } from './effects';

export { State } from './state';
export * from './reducer';
export { Actions };
export * from './selectors';
