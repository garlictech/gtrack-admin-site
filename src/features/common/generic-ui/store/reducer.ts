import * as ApiActions from './actions';
import { State } from './state';

export const initialState: State = {
  progressSpinnerOn: false
};

export type Action = ApiActions.AllActions;

// tslint:disable-next-line:only-arrow-functions
export function reducer(state = initialState, action: Action): State {
  switch (action.type) {
    case ApiActions.ActionTypes.ShowProgressSpinner:
      return { ...state, progressSpinnerOn: true, progressSpinnerText: action.textLabel };

    case ApiActions.ActionTypes.HideProgressSpinner:
      return { ...state, progressSpinnerOn: false, progressSpinnerText: undefined };

    default:
      return state;
  }
}
