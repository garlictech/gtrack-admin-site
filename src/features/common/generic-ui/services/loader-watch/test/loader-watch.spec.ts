import { TestBed } from '@angular/core/testing';
import { createSelector, Store, StoreModule } from '@ngrx/store';
import * as _ from 'lodash';
import { LoaderWatchService } from '..';
import { reducer } from '../../../store/reducer';
import * as fromSelectors from '../../../store/selectors';
import { featureName } from '../../../store/state';

describe('LoaderWatch service test', () => {
  let service: LoaderWatchService;
  let store: Store<any>;
  let spy: any;
  const initialState: any = {};
  const reducerObj: any = {};

  reducerObj[featureName] = reducer;
  const selector = createSelector(
    fromSelectors.selectFeature,
    (state: any) => state.foo
  );

  const initTestBed = _initialState => {
    TestBed.configureTestingModule({
      imports: [StoreModule.forRoot(reducerObj, { initialState: _initialState })]
    });

    store = TestBed.get(Store);
    spy = jest.spyOn(store, 'dispatch');
    service = new LoaderWatchService(store);
  };

  const runner = (_initialState, done, label?) => {
    initTestBed(_initialState);
    service.spinnerOnWorking$(selector, label).subscribe(() => {
      const call = spy.mock.calls[0][0];
      expect(call).toMatchSnapshot();
      done();
    });
  };

  it('should switch on the spinner if the selector emits true', done => {
    initialState[featureName] = { foo: true };
    runner(initialState, done, 'foobarlabel');
  });

  it('should switch off the spinner if the selector emits false', done => {
    initialState[featureName] = { foo: false };
    runner(initialState, done, 'foobarlabel');
  });

  it('should handle the undefined text label', done => {
    initialState[featureName] = { foo: true };
    runner(initialState, done);
  });
});
