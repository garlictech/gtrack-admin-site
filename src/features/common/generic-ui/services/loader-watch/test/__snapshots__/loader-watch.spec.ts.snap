// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoaderWatch service test should handle the undefined text label 1`] = `
ShowProgressSpinner {
  "textLabel": undefined,
  "type": "[UI] Show progress spinner",
}
`;

exports[`LoaderWatch service test should switch off the spinner if the selector emits false 1`] = `
HideProgressSpinner {
  "type": "[UI] Hide progress spinner",
}
`;

exports[`LoaderWatch service test should switch on the spinner if the selector emits true 1`] = `
ShowProgressSpinner {
  "textLabel": "foobarlabel",
  "type": "[UI] Show progress spinner",
}
`;
